### ISSUE TITLE
Loader with custom plugin results in "plugin.load is not a function"

### ISSUE THREAD

**Issue by fragsalat** [aurelia/cli#609]
**I'm submitting a bug report**

* **Library Version:**
0.28.0

**Please tell us about your environment:**
* **Operating System:**
Ubuntu 16.04

* **Node Version:**
6.3.1

* **NPM Version:**
4.2.0

* **Browser:**
all

* **Language:**
ESNext

**Current behavior:**
Adding a custom plugin to the aurelia-loader-default causes requirejs to throw the exception "TypeError: plugin.load is not a function"

**Expected/desired behavior:**
When adding a custom plugin to the default loader and requiring a resource with this plugin from the html file, the fetch function of the loader implementation is called.

**Steps to reproduce**
1. Create a new project using `au new loader-plugin-test`
    **Name**: loader-plugin-test
    **Platform**: Web
    **Transpiler**: Babel
    **Markup Processor**: None
    **CSS Processor**: Sass
    **Unit Test Runner**: Karma
    **Editor**: WebStorm
2. Create a new javascript file `src/loader-plugin/index.js`
```js
export function configure(config) {
    const loader = config.aurelia.loader;
    loader.addPlugin('sample', {
        fetch(address) {
            console.log('should be called');
            return loader.loadModule(address);
        }
    })
}
```
3. Load the feature in the main.js
```js
  aurelia.use
    .standardConfiguration()
    .feature('loader-plugin')
    .feature('resources');
```
4. Run the application `au run`
5. Open the application in the browser. There should be an exception in require now.

**Comment by JeroenVinke** [comment]
Hi, could you explain the use case?

**Comment by fragsalat** [comment]
Yes. We build in our Styleguide components to reduce redundancy over multiple applications with React/Preact because Webcomponents doesn't work out with Firefox. Therefore we want to load React components in Aurelia like in brian smith's aurelia-react-loader plugin. This works well for the Jspm setup but not via the Cli app

**Comment by JeroenVinke** [comment]
Thanks, I'll take a look

**Comment by fragsalat** [comment]
Thank you. I couldn't find the Problem or the solution in the short time I had.

**Comment by fragsalat** [comment]
Any news here?

**Comment by JeroenVinke** [comment]
Yes, it has to do with support for multiple plugin loaders. For example when you do `<require from="my-plugin!./some-custom-element"></require>` Aurelia will, in addition to `./some-custom-element.js`, try to load `my-plugin!./some-custom-element.html`. It does this with the `template-registry-entry` loader as `template-registry-entry!my-plugin!./some-custom-element.html`. This is where it fails, but I'll have to investigate where, which will take some time.

**Comment by fragsalat** [comment]
No worries I have some day's time :+1:  Thx for your time :)

**Comment by jmvtrinidad** [comment]
These error also appears when creating a plugin then that plugin requires to register in `globalResources` a customElement, see this [introduction](https://tochoromero.github.io/aurelia-table/#introduction) on **Aurelia CLI installation** part. Where it needs to add the html **manually** inside resources. In the behavior of **feature** installing a custom element does not require adding html as a dependecies it already locates it. In case of **CLI**, is it possible to include it automatically all html that has the same name with ts/js when using `au import/install`?

**Comment by fragsalat** [comment]
Maybe you should introduce a different sign to identify aurelia loader plugins^^

**Comment by JeroenVinke** [comment]
While I managed to get the custom loader to work for javascript files (by stripping the file extension so requirejs would find it in the bundle), it will take a longer to achieve your bigger goal (rendering react through a custom loader). There are a few bugs we need to take care of first: 
- https://github.com/aurelia/cli/issues/607
- https://github.com/aurelia/cli/issues/533 (fix for this is ready and will be merged soon)
- https://github.com/aurelia/cli/issues/583

I aim to get these bugs fixed as soon as possible, but I can't give you any timeline

**Comment by fragsalat** [comment]
Hey there. How did you strip the extension to get the custom loader working?
Maybe I can help in fixing the other things to get it work.

We'll need this in some weeks and it would be cool if we could use the cli tool instead of the skeletons :)

**Comment by JeroenVinke** [comment]
I used the following:

```
export function configure(config) {
    const loader = config.aurelia.loader;
    loader.addPlugin('loader-plugin', {
        fetch(address) {
            return loader.loadModule(address.replace('.js', ''))
            .then(x => { console.log(x); return x });
        }
    })
}
```

You're welcome to look into this, although it may be difficult to get it working in just a few weeks. Let me know if you get stuck.

**Comment by fragsalat** [comment]
This is working for you? In my test application I never reach the fetch function.

**Comment by JeroenVinke** [comment]
@fragsalat could you try [this app](https://github.com/JeroenVinke/aurelia-cli-609)?

**Comment by fragsalat** [comment]
Ah I saw the difference.

Before the require syntax was like <path>!<plugin> but now the direction switched.
old
```html
<require from="my-react-component!loader-plugin"></require>
```
new
```html
<require from="loader-plugin!my-react-component.js"></require>
```

Now I got the same error like you

**Comment by fragsalat** [comment]
So thanks to you I managed it to make it working but I had several problems. 
For instance there is a module called "prop-types" which is used by React / Preact which is written for nodejs environment. So the process variable I could mock with a script injected even before requirejs but the real problem was that module.exports wasn't there. So instead of having the real prop-types module I created a simple prop-types.js which just exports empty functions. Now the loader works even without the need to strip the extension.

```js
export function configure(config) {
    const loader = config.aurelia.loader;
    loader.addPlugin('styleguide-component', {
        fetch(address) {
            return loader.loadModule(address)
                .then(getComponents);
        }
    });
}
```

If you have detailed questions to how I did that now feel free to ask :)
